README.md
# EC2/RDS Scheduler

An improved version of AWS EC2 Scheduler](https://aws.amazon.com/answers/ec2-scheduler) that enables customers to easily configure custom start and stop schedules for their Amazon EC2 and RDS dev/test instances. The solution is easy to deploy and can help reduce operational costs for both development and production environments.

# New Features!

  -  Start/Stop RDS instance (http://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_StopInstance.html)
  -  Customzied Timezone support (pytz)
  -  Customzied AWS Region(s)
  -  Support start/stop EC2/RDS based on month day, nth weekdays, weekdays
  -  24x7, 24x5 scheduled tag support
  -  Default setting specified in AWS CloudWatch Event Target Input instead of DynamoDB table
  -  (bug fix) Support EC2 instance with encrypted EBS volumes

# Code 

## Cloudformation templates
- cform/ec2-scheduler.template

## Lambda source code
- code/ec2-scheduler.py

# Installation

You can deploy EC2/RDS scheduler either using AWS CLI (Linux) or from AWS Management Console:

## AWS CloudFormation Management Console
- CloudFormation Template:  cfrom/ec2-scheduler.template
- Paramters (please refer to https://aws.amazon.com/answers/infrastructure-management/ec2-scheduler/)

| Parameter | Default | Notes|
| ------ | ------ |------|
|DynamoDBTableName |  | Not Applicable, removed (default scheudler paramters are moved to CloudWatch Target Input)|
|ReadCapacityUnits |  | Same as above|
|WriteCapacityUnit |  | Same as above|
|Regions | all | AWS regions seperated by space(s) where EC2/RDS scheduler operates|
|DefaultTimeZone | utc | Default time zone for the scheduler|
|RDSSupport | Yes| Change to 'No' to disable RDS instance stop/start support|
|CustomRDSTagName | scheduler:rdcs-startstop| This tag identifies RDS instances to receive automated actions|
|S3BucketName | | S3 Bucket name where Lambda zipfile sits|

## AWS CLI

- Update stack parameters in deploy.sh 

      StackName="ec2-scheduler"
      DefaultTimeZone="Australia/Melbourne"
      RDSSupport="Yes"
      Schedule="5minutes"
      #Regions='ap-southeast-2 ap-southeast-1'
      Regions="ap-southeast-2"

- Deploy the stack using Shell script

```sh

./deploy.sh s3_bucket_name [profilename]

```

# Notes

## Dependency
- pytz : https://pypi.python.org/pypi/pytz
    
## Exceptions
Please install boto3  locally if you experience blow error message which is caused by the outdated boto3 library provided by AWS lambda.
- 'RDS' object has no attribute 'start_db_instance'
-  'RDS' object has no attribute 'stop_db_instance'

```sh
    cd code
    pip install boto3 -t .
    pip install pytz -t .
    zip -r -9 ../ec2-scheduler.zip *
    aws s3 cp ../ec2-scheduler.zip s3bucket
````


## Customzied TimeZone Support

Please refer to https://pypi.python.org/pypi/pytz for valid timezone (same as timezone supported by Amazon Linux).
The easiest way is to check folder /usr/share/zoneinfo of Amazon Linux, for example:

```sh
    cd /usr/share/zoneinfo
    find . -type f |sed s'/^\.\///'|grep Australia
    ....
    Australia/Melbourne
    ...
```

"Australia/Melbourne" is a valid timezone.

## IAM Role RDS Permissions
AWS accounced RDS instance stop/stop support in June 1, 2017, as this file is wriiten, IAM role allow unstricted RDS API access "rds:*" because no AWS document states which RDS api applies to rds instance stop and start operations.
URL:

    - http://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAM.ResourcePermissions.html
    - https://aws.amazon.com/about-aws/whats-new/2017/06/amazon-rds-supports-stopping-and-starting-of-database-instances/

# Applying Custom Parameters
You can apply custom start and stop parameters to an EC2 or RDS instance which will orverwide the default values you set during initial deployment. To do this, mofidy the tag value to specify the alternative settings.

The EC2/RDS Scheduler will read tag values, looking for four possible custom parameters in foloowing order: 
<start time>; <stop time>; <time zone>; <active days(s)>

You can sepereate each values with a semicolon or colon. The following table gives acceptable input values for each field

|Tag Value Field|Acceptable input values|Scheduled Action/Note |
| ------ | ------ | ------ | 
|start time | none | No action|
|| 24x7| Start the instance at any time if it is stopped|
|| default| The instance will start and stop on the default schedule|
|| true| The instance will start and stop on the default schedule|
|| HHMM| Time in 24-hour format (Default time zone or timezone specified, with no colon)|
|stop time | none | No action|
|| HHMM| Time in 24-hour format (Default time zone or timezone specified, with no colon)|
|time zone| <empty>| Default scheduler time zone |
||utc| UTC time zone |
||Australia/Melbourne| Or any pytz library supported time zone value |
|active day(s)|all| All days |
||weekdays| From Monday to Friday |
||sat,1,2| Saturday, 1st and 2nd in each month|
||sat/2, fri/4| The second Saturday and fourth Friday in each month|

# Example Tag Value

The following table gives examples of different tag values and the resulting Scheduler actions

|Example Tag Value|EC2 Scheduler Action|
|------ | ------|
|24x7 | start RDS/EC2 instance at any time if it is stopped|
|24x5 | start RDS/EC2 instance at DefaultStartTime on Monday, and stop at DefaultStopTime on Friday (DefaultTimeZone) |
|24x5;;Australia/Adelaide; | start at DefaultStartTime on Monday, and stop DefaultStopTime on Friday (Australia/Adelaide timezone)|
|none | No action|
|default | The instance will start and stop on the default schedule|
|true | The instance will start and stop on the default schedule|
|0800;;;weekdays |Start the instance at 08:00 (Default Timezone) in weekdays if it is stopped|
|;1700;;weekdays |Stop the instance at 17:00 (Default Timezone) in weekdays if it is running.|
|0800;1800;utc;all|The instance will start at 0800 hours and stop at 1800 hours on all days|
|0001;1800;Etc/GMT+1;Mon/1| The instance will start at 0001 hour and stop at 1800 hour (first Monday of every month, Etc/GMT+1 timezone)|
|1000;1700;utc;weekdays | The instance will start at 1000 hours and stop at 1700 hours Monday through Friday.|
|1030;1700;utc;mon,tue,fri| The instance will start at 1030 hours and stop at 1700 hours on Monday, Tuesday, and Friday only.|
|1030;1700;utc;mon,tue,fri,1,3 |The instance will start at 1030 hours and stop at 1700 hours on Monday, Tuesday, and Friday or date 1,3 only.|
|1030;1700;utc;1 |The instance will start at 1030 hours and stop at 1700 hours on date 1 only.|
|1030;1700;utc;01,fri | The instance will start at 1030 hours and stop at 1700 hours on date 1 and Friday.|
|0815;1745;utc;wed,thu |The instance will start at 0815 hours and stop at 1745 hours on Wednesday and Thursday.|
|none;1800;utc;weekdays| The instance stop at 1800 hours Monday through Friday. |
|0800;none;utc;weekdays| The instance start at 0800 hours Monday through Friday. |
|1030;1700;utc;mon,tue,fri,1,3,sat/1|The instance will start at 1030 hours and stop at 1700 hours on Monday, Tuesday, and Friday ,date 1,3 or the first Saturday in every month (utc TimeZone)|
|1030;1700;;mon,tue,fri,1,3,sat/1|The instance will start at 1030 hours and stop at 1700 hours on Monday, Tuesday, and Friday ,date 1,3 or the first Saturday in every month (Default TimeZone)|
|1030;1700;Australia/Sydney;mon,tue,fri,1,3,sat/1 | The instance will start at 1030 hours and stop at 1700 hours on Monday, Tuesday, and Friday ,date 1,3 or the first Saturday in every month (Australia/Sydney TimeZone)|

# Author
- Initial version: AWS provided
- Updated by:
Eric Ho (eric.ho@datacom.com.au, hbwork@gmail.com, https://www.linkedin.com/in/hbwork/)
Last update: October 11, 2017

***

Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.

Licensed under the Amazon Software License (the "License"). You may not use this file except in compliance with the License. A copy of the License is located at

    http://aws.amazon.com/asl/

or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGES.txt
Author: Eric Ho (eric.ho@datacom.com.au)
Date: May 29, 2017
Changes:
- code/ec2-scheduler.py: month-days and/or weekdays are supported in EC2 instance tags
  for example:
  2230;0725;utc;mon,fri – every Monday and Friday
  2230;0725;utc;weekdays – every weekday from Monday to Friday
  none;1700;utc;mon,fri,1,31  - every Monday, Friday, or 1st or 31st every month
- cform/ec2-scheduler.template.json :
  Add "kms.CreateGrant" permission to IAM role "ec2SchedulerRole" to fix issues of starting EC2 instance with encrypted EBS volume
  Modify S3 bucket name to dcp-install for customized lambda function code

Date: June 2, 2017
Changes:
- code/ec2-scheduler.py: nth weekdays are supported in EC2 instance tags
  for example:
  2230;0725;utc;mon/1
  2230;0725;utc;weekdays,sat/1,sun/4
  none;1700;utc;mon,fri,1,31

Date: June 8, 2017
Changes:
- cform/ec2-scheduler.template: new Parameter DefaultTimeZone
- code/ec2-scheduler.py: customized timezone support (http://pytz.sourceforge.net/)
  0830;1750;;all  Start instance at 08:30 and stop at 17:30 all days (DefaultTimeZone)
  0830;1750;Australia/Sydney;all Start instance at 08:30 and stop at 17:30 all days (Australia/Sydney time zone)
  0830;1750;utc;all Start instance at 08:30 and stop at 17:30 all days (UTC timezone)
  0830;1750;Etc/GMT+10;all Start instance at 08:30 and stop at 17:30 all days (Etc/GMT+10 timezone)
  0830;1750;Asia/Tokyo;all Start instance at 08:30 and stop at 17:30 all days (Asia/Tykyo)

Date: July 3, 2017
Changes:
- code/ec2-scheduler.py: bug fix - fails to start instance if startTime is midnight 00xx 

Date: July 6, 2017
Changes:
- code/ec2-scheduler.py: 
   bug fix: nthweek REGEX pattern does match /[2-5] in nthweekdays 
   enhancement: support ":" and ";" as separtor in the tag value

Date: July 19,2017
Changes:
- New feature: RDS instance support
- cform/ec2-scheduler.template.json :
  Remove DynamoDB and add default parameters to CloudWatch rule event input
  IAM role - remove dynamoDB access and allow RDS access (to be customzied, allows rds:* for the moment)
  Additional parameters: Regions, CustomRDSTagName, S3BucketName
  Lambda function: memory size 256MB
- code/ec2-scheduler.py: 
  Read default parameter from Cloudwatch event input
  customized AWS Region(s)
  startTime 24x7 - start instance at any time if it is stopped
  RDS instance support
  Re-structure python code 
- deploy.sh
  Shell script to create/update CloudFormation stack (Jenkins Job script)

Date: October 11,2017
Changes:
- Enhancement: 24x5 support - start EC2/RDS Instance at DefaultStartTime on Monday, and stop  EC2/RDS instance at DefaultStopTime on Friday

Note about RDS support:
- no CloudWatch Matrics
- no anonymous data sent to AWS
- http://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_StopInstance.html

How to deploy:
  Deploy CloudFromation stack from CF template cform\ec2-scheduler.template by following document https://s3.amazonaws.com/solutions-reference/ec2-scheduler/latest/ec2-scheduler.pdf or http://docs.aws.amazon.com/solutions/latest/ec2-scheduler/deployment.html

Note: 
  - python code was uploaded to s3://dcp-install
  - nth weekdays format: weekday/nthweek
    weekday(case insensitive): mon, tue,wed,thu,fri,sat,sun 
    nthweek: 1 (first week in every month),2 (second) week, 3 (third week),4(fourth week)




deploy.sh
#!/bin/bash
# 
# deploy.sh
#   Shell script to deploy Aws EC2/RDS Scheduer 
# Dependecy: 
#   - pip
#   - AWS CLI configured with credential and default AWS regions
# Reference: https://aws.amazon.com/answers/infrastructure-management/ec2-scheduler/
# Author: Eric Ho (eric.ho@datacom.com.au, hbwork@gmail.com, https://www.linkedin.com/in/hbwork/)
#

function assume_role {
    if [ $(date "+%s") -ge ${TokenExpiration} ]
    then
        unset AWS_ACCESS_KEY_ID
        unset AWS_SECRET_ACCESS_KEY
        unset AWS_SECURITY_TOKEN
        unset TokenExpiration

        echo "(Renew) Assuming role arn:aws:iam::${AccountId}:role/${RoleName}"

        #Default session expiration period is 1 hour
        aws sts assume-role --role-session-name JenkinCiSession \
            --role-arn arn:aws:iam::${AccountId}:role/${RoleName} \
            --external-id ${ExternalId}  \
            | grep -w 'AccessKeyId\|SecretAccessKey\|SessionToken' \
            | awk  '{print $2}' \
            | sed  's/"//g;s/,//'\
            > awscre

        export AWS_ACCESS_KEY_ID=`sed -n '3p' /tmp/awscre`
        export AWS_SECRET_ACCESS_KEY=`sed -n '1p' /tmp/awscre`;
        export AWS_SECURITY_TOKEN=`sed -n '2p' /tmp/awscre`

        # Sesson expiration time  (50 minutes)
        export TokenExpiration=$(expr $(date "+%s") + 3000)

        rm awscre
    fi
}

# assume_role - Deploy EC2 Scheduer by assuming IAM role in other AWS account

#if [ "$#" -ne 5 ] ; then
#  echo "Usage: ./deploy.sh s3_bucket_name AwsAccountId IamRoleNamee externalId"
#  exit 1
#fi

#export AccountId=${2}
#export RoleName=${3}
#export ExternalId=${4}
#export TokenExpiration=$(date "+%s")
#assume_role

# End of assume_role


if [ "$#" -lt 1 ] ; then
    echo "Usage: ./deploy.sh s3_bucket_name [profilename]"
    exit 1
fi

#S3 BucketName for zip file
S3BucketName=${1}

# AWS CLI default profile 
ProfileName=${2:-default}

#Stack Deploy Parameters
StackName="ec2-scheduler"
DefaultTimeZone="Australia/Melbourne"
RDSSupport="Yes"
Schedule="5minutes" 
#Regions='ap-southeast-2 ap-southeast-1'
Regions="ap-southeast-2"

# End of Customzation 

rm -f *.zip

cd code

# boto3 library provided in lambda does not support rds.start_db_instance and rds.stop_db_instance (in July 20, 2017)
pip install boto3 -t .
pip install pytz -t .

zip -r -9 ../ec2-scheduler.zip *

aws s3 cp ../ec2*.zip s3://${S3BucketName}

cd ..

aws cloudformation validate-template --template-body file://cform/ec2-scheduler.template

set +e

StackStatus=$(aws cloudformation describe-stacks --stack-name ${StackName} --query Stacks[0].StackStatus --output text)

set -e

if [ ${#StackStatus} -eq 0 ]
then
    echo "Please ignore the Validation Error messagege above. Create CloudFormation stack ${StackName} ..."

    aws cloudformation create-stack --stack-name ${StackName} \
		  --template-body file://cform/ec2-scheduler.template \
    	--capabilities CAPABILITY_IAM CAPABILITY_NAMED_IAM \
    	--profile ${ProfileName} \
    	--parameters \
    	  ParameterKey=DefaultTimeZone,ParameterValue=${DefaultTimeZone}  \
    	  ParameterKey=RDSSupport,ParameterValue=${RDSSupport} \
    	  ParameterKey=Schedule,ParameterValue=${Schedule} \
    	  ParameterKey=S3BucketName,ParameterValue=${S3BucketName} \
    	  ParameterKey=Regions,ParameterValue="${Regions}"

    sleep 30

elif [ ${StackStatus} == 'CREATE_COMPLETE' -o ${StackStatus} == 'UPDATE_COMPLETE' ]
then
    echo "Update stack ${StackName} ..."

    ChangeSetName="${StackName}-$(uuidgen)"

    aws cloudformation create-change-set --stack-name ${StackName} \
          --template-body file://cform/ec2-scheduler.template \
    	    --profile ${ProfileName} \
          --change-set-name ${ChangeSetName} \
          --capabilities CAPABILITY_IAM CAPABILITY_NAMED_IAM \
    	    --parameters \
    	      ParameterKey=DefaultTimeZone,ParameterValue=${DefaultTimeZone}  \
    	      ParameterKey=RDSSupport,ParameterValue=${RDSSupport} \
    	      ParameterKey=Schedule,ParameterValue=${Schedule} \
    	      ParameterKey=S3BucketName,ParameterValue=${S3BucketName} \
    	      ParameterKey=Regions,ParameterValue="${Regions}"
    sleep 10

    ChangeSetStatus=$(aws cloudformation describe-change-set  \
    	  --profile ${ProfileName} \
        --change-set-name ${ChangeSetName} \
        --stack-name ${StackName} \
        --query Status \
        --output text)

    #CREATE_IN_PROGRESS , CREATE_COMPLETE , or FAILED
    while [ ${ChangeSetStatus} == 'CREATE_IN_PROGRES' ]
    do
        sleep 30
        ChangeSetStatus=$(aws cloudformation describe-change-set \
            --change-set-name ${ChangeSetName} \
    	      --profile ${ProfileName} \
            --stack-name ${StackName} \
            --query Status \
            --output text)
    done

	  if [ ${ChangeSetStatus} == 'FAILED' ]
	  then
        echo "Update lambda code using zip file"
        FunctionName=$(aws lambda list-functions --query Functions[].FunctionName \
    	      --profile ${ProfileName} \
            | grep ec2SchedulerOptIn |sed s/\",*//g |tr -d '[:space:]')

        aws lambda update-function-code --profile ${ProfileName} --function-name ${FunctionName} --zip-file fileb://ec2-scheduler.zip
    else
        echo "Execute change set ${ChangeSetName}"
        aws cloudformation execute-change-set --change-set-name ${ChangeSetName} --stack-name ${StackName} --profile ${ProfileName}
        sleep 30
    fi
else
    echo "Failed to create or update stack ${StackName} (Unexpected stack status: ${StackStatus})"
    exit 1
fi

StackStatus=$(aws cloudformation describe-stacks \
    --stack-name ${StackName} \
    --query Stacks[0].StackStatus \
    --output text  \
    --profile ${ProfileName})

#CREATE_IN_PROGRESS
#UPDATE_IN_PROGRESS

while [ $StackStatus == "CREATE_IN_PROGRESS" -o $StackStatus == "UPDATE_IN_PROGRESS" ]
do
    sleep 30
    StackStatus=$(aws cloudformation describe-stacks \
      --stack-name ${StackName} \
      --query Stacks[0].StackStatus \
      --output text  \
      --profile ${ProfileName})
done

# CREATE_COMPLETE
# UPDATE_COMPLETE
# UPDATE_COMPLETE_CLEANUP_IN_PROGRESS
if [ $StackStatus != "CREATE_COMPLETE" -a $StackStatus != "UPDATE_COMPLETE_CLEANUP_IN_PROGRESS" -a $StackStatus != "UPDATE_COMPLETE" ]
then
    echo "Create/Update stack failed - ${StackStatus}"
    exit 1
fi

echo "Create/Update stack succeeded"
# __EOF__


code/ec2-scheduler.py
######################################################################################################################
#  Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           #
#                                                                                                                    #
#  Licensed under the Amazon Software License (the "License"). You may not use this file except in compliance        #
#  with the License. A copy of the License is located at                                                             #
#                                                                                                                    #
#      http://aws.amazon.com/asl/                                                                                    #
#                                                                                                                    #   
#  or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES #
#  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    #
#  and limitations under the License.                                                                                #
######################################################################################################################


import boto3
import datetime
import json
from urllib2 import Request
from urllib2 import urlopen
from collections import Counter

# Import re to analyse month day
import re

# Import pytz to support local timezone
import pytz


#Default startTime, stopTime, TimeZone
defaultStartTime = 'none'
defaultStopTime = 'none'
defaultTimeZone = 'utc'
defaultDaysActive = 'all'

def putCloudWatchMetric(region, instance_id, instance_state):
    
    cw = boto3.client('cloudwatch')

    cw.put_metric_data(
        Namespace='EC2Scheduler',
        MetricData=[{
            'MetricName': instance_id,
            'Value': instance_state,

            'Unit': 'Count',
            'Dimensions': [
                {
                    'Name': 'Region',
                    'Value': region
                }
            ]
        }]
        
    )

def scheduler_action(tagValue):
    # Weekdays Interpreter
    weekdays = ['mon', 'tue', 'wed', 'thu', 'fri']

    # Monthdays Interpreter (1..31)
    monthdays = re.compile(r'^(0?[1-9]|[12]\d|3[01])$')

    # nth weekdays Interpreter
    nthweekdays=re.compile('\w{3}/\d{1}')

    # Split out Tag & Set Variables to default
    default1 = 'default'
    default2 = 'true'

    #Default scheduler action
    Action = 'None'

    ptag = tagValue.replace(':',';').split(";")

    startTime = defaultStartTime
    stopTime = defaultStopTime
    timeZone = defaultTimeZone
    daysActive = defaultDaysActive

    # Default Timzone
    tz = pytz.timezone(defaultTimeZone)

    # Valid timezone
    isValidTimeZone = True

    # Parse tag-value
    if len(ptag) >= 1:
        if ptag[0].lower() in (default1, default2):
            startTime = defaultStartTime
        else:
            startTime = ptag[0]
            stopTime = ptag[0]

    # Support 24x7
    if startTime  == '24x7':
        return "START"

    if len(ptag) >= 2:
        stopTime = ptag[1]
    if len(ptag) >= 3:
        #Timezone is case senstive
        timeZone = ptag[2]

        # timeZone is not empty and not DefaultTimeZone
        if timeZone != defaultTimeZone and timeZone != '':
            # utc is not included in pytz.all_timezones
            if timeZone != 'utc':
                if timeZone in pytz.all_timezones:
                    tz = pytz.timezone(timeZone)
                # No action if timeZone is not supported
                else:
                    print "Invalid time zone :", timeZone
                    isValidTimeZone = False
            # utc timezone
            else:
                tz = pytz.timezone('utc')

    if len(ptag) >= 4:
        daysActive = ptag[3].lower()

    now = datetime.datetime.now(tz).strftime("%H%M")

    if  datetime.datetime.now(tz).strftime("%H") != '00':
        nowMax = datetime.datetime.now(tz) - datetime.timedelta(minutes=59)
        nowMax = nowMax.strftime("%H%M")
    else:
        nowMax = "0000"

    nowDay = datetime.datetime.now(tz).strftime("%a").lower()
    nowDate = int(datetime.datetime.now(tz).strftime("%d"))

    isActiveDay = False

    # 24x5 support 
    if startTime  == '24x5':
       if nowDay == 'mon':
           startTime = defaultStartTime
           stopTime = 'none'
           isActiveDay = True
       elif nowDay == 'fri':
           isActiveDay = True
           startTime = 'none'
           stopTime = defaultStopTime

    elif daysActive == "all":
        isActiveDay = True
    elif daysActive == "weekdays":
        if (nowDay in weekdays):
            isActiveDay = True
    else:
        for d in daysActive.split(","):
            # mon, tue,wed,thu,fri,sat,sun ?
            if d.lower() == nowDay:
                    isActiveDay = True
            # Month days?
            elif monthdays.match(d):
                if int(d) == nowDate:
                     isActiveDay = True
            # mon/1 first Monday of the month
            # tue/2 second Tuesday of the month
            # Fri/3 third Friday of the month
            # sat/4 forth Saturday of the month
            elif nthweekdays.match(d):
                (weekday,nthweek) = d.split("/")

                if (weekday.lower() == nowDay) and ( nowDate >= (int(nthweek) * 7 - 6)) and (nowDate <= (int(nthweek) * 7)):
                   isActiveDay = True

    if startTime >= str(nowMax) and startTime <= str(now) and isActiveDay == True and isValidTimeZone == True:
        Action = "START"

    if stopTime >= str(nowMax) and stopTime <= str(now) and isActiveDay == True and isValidTimeZone == True:
        Action = "STOP"

    return Action

def lambda_handler(event, context):
    global defaultStartTime
    global defaultStopTime
    global defaultTimeZone
    global defaultDaysActive

    print "Running EC2 and RDS Scheduler"

    ec2 = boto3.client('ec2')

    if event['Regions'] == 'all':
        AwsRegionNames = []
        for r in ec2.describe_regions()['Regions']:
            AwsRegionNames.append(r['RegionName'])
    else:
        AwsRegionNames = event['Regions'].split()

    print "Operate in regions: ", " ".join(AwsRegionNames)

    # Reading Default Values from CloudWatch Rule Input event
    customTagName = event['CustomTagName']
    customTagLen = len(customTagName)
    sendData = event['SendAnonymousData'].lower()
    createMetrics = event['CloudWatchMetrics'].lower()
    UUID = event['UUID']

    # Set global default value
    defaultStartTime = event['DefaultStartTime']
    defaultStopTime = event['DefaultStopTime']
    defaultTimeZone = event['DefaultTimeZone']
    defaultDaysActive = event['DefaultDaysActive']

    #Customized RDS Tag Name
    RDSSupport = event['RDSSupport']
    customRDSTagName = event['CustomRDSTagName']
    customRDSTagLen = len(customRDSTagName)

    TimeNow = datetime.datetime.utcnow().isoformat()
    TimeStamp = str(TimeNow)

    # Declare Dicts
    regionDict = {}
    allRegionDict = {}
    regionsLabelDict = {}
    postDict = {}

    for region_name in AwsRegionNames:
        try:
            # Create connection to the EC2 using Boto3 resources interface
            ec2 = boto3.resource('ec2', region_name = region_name)

            # Declare Lists
            startList = []
            stopList = []
            runningStateList = []
            stoppedStateList = []

            # List all instances
            instances = ec2.instances.all()

            print "Creating ", region_name, " EC2 instance lists..."

            for i in instances:
                if i.tags != None:
                    for t in i.tags:
                        if t['Key'][:customTagLen] == customTagName:

                            state = i.state['Name']
                            itype = i.instance_type

                            # Post current state of the instances
                            if createMetrics == 'enabled':
                                if state == "running":
                                    putCloudWatchMetric(region_name, i.instance_id, 1)
                                if state == "stopped":
                                    putCloudWatchMetric(region_name, i.instance_id, 0)

                            action = scheduler_action(tagValue = t['Value'])

                            # Append to start list
                            if action == "START" and state == "stopped":
                                if i.instance_id not in startList:
                                    startList.append(i.instance_id)
                                    print i.instance_id, " added to START list"
                                    if createMetrics == 'enabled':
                                        putCloudWatchMetric(region_name, i.instance_id, 1)
                                # Instance Id already in startList

                            # Append to stop list
                            if action == "STOP" and state == "running":
                                if i.instance_id not in stopList:
                                    stopList.append(i.instance_id)
                                    print i.instance_id, " added to STOP list"
                                    if createMetrics == 'enabled':
                                        putCloudWatchMetric(region_name, i.instance_id, 0)
                                # Instance Id already in stopList

                            if state == 'running':
                                runningStateList.append(itype)
                            if state == 'stopped':
                                stoppedStateList.append(itype)

            # Execute Start and Stop Commands
            if startList:
                print "Starting", len(startList), "instances", startList
                ec2.instances.filter(InstanceIds=startList).start()
            else:
                print "No Instances to Start in region ",  region_name

            if stopList:
                print "Stopping", len(stopList) ,"instances", stopList
                ec2.instances.filter(InstanceIds=stopList).stop()
            else:
                print "No Instances to Stop in region ", region_name

            # Built payload for each region
            if sendData == "yes":
                countRunDict = {}
                typeRunDict = {}
                countStopDict = {}
                typeStopDict = {}
                runDictType = {}
                stopDictType = {}
                runDict = dict(Counter(runningStateList))
                for k, v in runDict.iteritems():
                    countRunDict['Count'] = v
                    typeRunDict[k] = countRunDict['Count']

                stopDict = dict(Counter(stoppedStateList))
                for k, v in stopDict.iteritems():
                    countStopDict['Count'] = v
                    typeStopDict[k] = countStopDict['Count']

                runDictType['instance_type'] = typeRunDict
                stopDictType['instance_type'] = typeStopDict

                typeStateSum = {}
                typeStateSum['running'] = runDictType
                typeStateSum['stopped'] = stopDictType
                StateSum = {}
                StateSum['instance_state'] = typeStateSum
                regionDict[region_name] = StateSum
                allRegionDict.update(regionDict)
            
        except Exception as e:
            print ("Exception: "+str(e))
            continue

        if RDSSupport == "Yes":
            # Declare Lists
            rdsStartList = []
            rdsStopList = []
            rdsRunningStateList = []
            rdsStoppedStateList = []

            print "Creating ", region_name, " RDS instance lists..."
            try:
                rds = boto3.client('rds', region_name =  region_name)
                rds_instances = rds.describe_db_instances()

                for rds_instance in rds_instances['DBInstances']:

                    if len(rds_instance['ReadReplicaDBInstanceIdentifiers']):
                        print "----- No action against RDS instance (with read replica): ", rds_instance['DBInstanceIdentifier']
                        continue

                    if "ReadReplicaSourceDBInstanceIdentifier" in rds_instance.keys():
                        print "----- No action against RDS instance (replication): ", rds_instance['DBInstanceIdentifier']
                        continue

                    if rds_instance['MultiAZ']:
                        print "----- No action against RDS instance (Multiple AZ): ", rds_instance['DBInstanceIdentifier']
                        continue

                    if rds_instance['DBInstanceStatus'] not in ["available","stopped"]:
                        print "----- No action against RDS instance (unsupported status):", rds_instance['DBInstanceIdentifier']
                        continue

                    # Query RDS instance tags 
                    response = rds.list_tags_for_resource( ResourceName = rds_instance['DBInstanceArn'])
                    tags = response['TagList']

                    for t in tags:
                        if t['Key'][:customRDSTagLen] == customRDSTagName:

                            action = scheduler_action(tagValue = t['Value'])
                            state = rds_instance['DBInstanceStatus'] 

                            # Append to start list
                            if action == "START" and state == "stopped":
                                if rds_instance['DBInstanceIdentifier'] not in rdsStartList:
                                    rdsStartList.append(rds_instance['DBInstanceIdentifier'])
                                    print rds_instance['DBInstanceIdentifier'], " added to RDS START list"
                                # Instance Id already in rdsStartList

                            # Append to stop list
                            if action == "STOP" and state == "available":
                                if rds_instance['DBInstanceIdentifier'] not in rdsStopList:
                                    rdsStopList.append(rds_instance['DBInstanceIdentifier'])
                                    print rds_instance['DBInstanceIdentifier'], " added to RDS STOP list"
                                # Instance Id already in rdsStopList

                # Execute Start and Stop Commands
                if rdsStartList:
                    print "Starting", len(rdsStartList), "RDS instances", rdsStartList
                    for DBInstanceIdentifier in rdsStartList:
                        rds.start_db_instance(DBInstanceIdentifier = DBInstanceIdentifier)
                else:
                    print "No RDS Instances to Start in region ",  region_name

                if rdsStopList:
                    print "Stopping", len(rdsStopList) ,"RDS instances", rdsStopList
                    for DBInstanceIdentifier in rdsStopList:
                        rds.stop_db_instance(DBInstanceIdentifier = DBInstanceIdentifier)
                else:
                    print "No RDS Instances to Stop in region ", region_name

            except Exception as e:
                print ("Exception: "+str(e))
                continue

    # Build payload for the account
    if sendData == "yes":
        regionsLabelDict['regions'] = allRegionDict
        postDict['Data'] = regionsLabelDict
        postDict['TimeStamp'] = TimeStamp
        postDict['Solution'] = 'SO0002'
        postDict['UUID'] = UUID
        # API Gateway URL to make HTTP POST call
        url = 'https://metrics.awssolutionsbuilder.com/generic'
        data=json.dumps(postDict)
        headers = {'content-type': 'application/json'}
        req = Request(url, data, headers)
        rsp = urlopen(req)
        content = rsp.read()
        rsp_code = rsp.getcode()
        print ('Response Code: {}'.format(rsp_code))

# Local version
if  __name__ =='__main__':
    event = {
        "DefaultStartTime": "0800",
        "DefaultStopTime": "1800",
        "UUID": "27b0ceaa-2162-47d0-9ed1-e7e76f22cbba",
        "SolutionName": "EC2Scheduler",
        "DefaultDaysActive": "all",
        "Regions": "ap-southeast-2",
        "DefaultTimeZone": "Australia/Melbourne",
        "RDSSupport": "Yes",
        "SendAnonymousData": "No",
        "CustomTagName": "scheduler:ec2-startstop",
        "CustomRDSTagName": "scheduler:rds-startstop",
        "CloudWatchMetrics": "Enabled"
    }

    context = None

    lambda_handler(event = event, context = context)

#EOF
